### PR TITLE
[#545][FOLLOWUP]feat(operator): support specifying custom tolerations

### PR DESCRIPTION
**Pull request by crain-cn**
### What changes were proposed in this pull request?

Support coordinator/shuffler server's  tolerations to the fields of RSS spec

### Why are the changes needed?
1. Production environment deployment needs to use.
2. In order to change parameters flexibly.

### Does this PR introduce _any_ user-facing change?
For RSS cluster admin, they can set custom tolerations for shuffle servers and coordinators.

### How was this patch tested?
Manually verified.
image

### CODE REVIEW

**Comment by zuston** [comment]
Could you help review this? @wangao1236 @advancedxy

**Comment by kaijchen** [comment]
Thanks @crain-cn for the work. Please take a look at the test failures in CI.

**Comment by crain-cn** [comment]
> Thanks @crain-cn for the work. Please take a look at the test failures in CI.

Thanks @kaijchen

File: deploy/kubernetes/operator/api/uniffle/v1alpha1/remoteshuffleservice_types.go [hunk 1]
```diff
@@ -249,6 +253,15 @@ type MainContainer struct {
...
 	VolumeMounts []corev1.VolumeMount `json:"volumeMounts,omitempty"`
+
+	// NodeSelector is a selector which must be true for the pod to fit on a node.
+	// Selector which must match a node's labels for the pod to be scheduled on that node.
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// Indicates the tolerations the pods under this subset have.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
```
[the remark this answers is not in the document]
Comment (crain-cn) [reply deploy/kubernetes/operator/api/uniffle/v1alpha1/remoteshuffleservice_types.go]: The fields in MainContainer have been removed.

File: deploy/kubernetes/operator/api/uniffle/v1alpha1/remoteshuffleservice_types.go [hunk 2]
```diff
@@ -218,6 +218,10 @@ type RSSPodSpec struct {
 	// Selector which must match a node's labels for the pod to be scheduled on that node.
 	// +optional
 	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// Indicates the tolerations the pods under this subset have.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
```
[the remark this answers is not in the document]
Comment (crain-cn) [reply deploy/kubernetes/operator/api/uniffle/v1alpha1/remoteshuffleservice_types.go:224]: No need to add affinity, toleration configuration needs to be added in kubernetes 1.21.3 for our large-scale services, with many taints.

File: deploy/kubernetes/operator/pkg/controller/sync/coordinator/coordinator.go [hunk 3]
```diff
@@ -178,12 +178,7 @@ func GenerateDeploy(rss *unifflev1alpha1.RemoteShuffleService, index int) *appsv
 	podSpec := corev1.PodSpec{
 		HostNetwork:        *rss.Spec.Coordinator.HostNetwork,
 		ServiceAccountName: utils.GenerateCoordinatorName(rss),
-		Tolerations: []corev1.Toleration{
-			{
-				Effect: corev1.TaintEffectNoSchedule,
-				Key:    "node-role.kubernetes.io/master",
-			},
-		},
+		Tolerations:  rss.Spec.Coordinator.Tolerations,
```
[the remark this answers is not in the document]
Comment (crain-cn) [reply deploy/kubernetes/operator/pkg/controller/sync/coordinator/coordinator.go]: If using the default configuration, it is definitely not supported. The operator must be modified. If not supported, we can only do secondary development here. We really hope to cooperate with you here. Our single node scale is above 4000 nodes, and we also hope that the community can provide help. Thank you!

[the remark this answers is not in the document]
Comment (crain-cn) [reply deploy/kubernetes/operator/pkg/controller/sync/coordinator/coordinator.go]: Real business scenarios should not be deployed to the master rule node, especially for large single-node scales like ours, to consider the isolation of the business and ensure the stability of the business.

[the remark this answers is not in the document]
Comment (crain-cn) [reply deploy/kubernetes/operator/pkg/controller/sync/coordinator/coordinator.go]: If rss.Spec.Coordinator.Tolerations is not specified, can I modify the code to default to the original configuration?

File: deploy/kubernetes/operator/api/uniffle/v1alpha1/remoteshuffleservice_types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (crain-cn) [reply deploy/kubernetes/operator/api/uniffle/v1alpha1/remoteshuffleservice_types.go:224]: I can support this together and submit a pull request.

**Comment by jerqi** [comment]
@crain-cn Could you add some uts for this pr?

File: deploy/kubernetes/operator/pkg/controller/sync/coordinator/coordinator.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (crain-cn) [reply deploy/kubernetes/operator/pkg/controller/sync/coordinator/coordinator.go]: > what do you mean by large **singe-node** scales?
A cluster with many nodes. big cluster.

File: deploy/kubernetes/operator/pkg/controller/sync/coordinator/coordinator.go [hunk 4]
```diff
@@ -178,14 +178,9 @@ func GenerateDeploy(rss *unifflev1alpha1.RemoteShuffleService, index int) *appsv
...
-				Key:    "node-role.kubernetes.io/master",
-			},
-		},
-		Volumes:      rss.Spec.Coordinator.Volumes,
-		NodeSelector: rss.Spec.Coordinator.NodeSelector,
+		Tolerations:        rss.Spec.Coordinator.Tolerations,
+		Volumes:            rss.Spec.Coordinator.Volumes,
+		NodeSelector:       rss.Spec.Coordinator.NodeSelector,
 	}
 	configurationVolume := corev1.Volume{
```
[the remark this answers is not in the document]
Comment (crain-cn) [reply deploy/kubernetes/operator/pkg/controller/sync/coordinator/coordinator.go:185]: Thanks 😄

**Comment by crain-cn** [comment]
> @crain-cn Could you add some uts for this pr?
Yeah.

File: deploy/kubernetes/operator/api/uniffle/v1alpha1/remoteshuffleservice_types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (crain-cn) [reply deploy/kubernetes/operator/api/uniffle/v1alpha1/remoteshuffleservice_types.go:224]: > 👍 , thanks.
Affinities, can I submit to the next pr?

**Comment by crain-cn** [comment]
@advancedxy Affinity fields , can I submit to the next pr?

**Comment by crain-cn** [comment]
> > @advancedxy Affinity fields , can I submit to the next pr?
> 
> of course.

Thanks ! 😄

**Comment by crain-cn** [comment]
> @crain-cn Could you add some uts for this pr?

Can I make it up later?

**Comment by crain-cn** [comment]
> @crain-cn Could you add some uts for this pr?
Already written.
#641 [0cfc745](https://github.com/apache/incubator-uniffle/pull/641/commits/0cfc74516ac781caf4eb18422294cc29fa222090)

**Comment by crain-cn** [comment]
> Cloud you add some UTs for this new field?

Already written.
https://github.com/apache/incubator-uniffle/pull/641 [0cfc745](https://github.com/apache/incubator-uniffle/pull/641/commits/0cfc74516ac781caf4eb18422294cc29fa222090)

**Comment by jerqi** [comment]
> > @crain-cn Could you add some uts for this pr?
> > Already written.
> > #641 [0cfc745](https://github.com/apache/incubator-uniffle/pull/641/commits/0cfc74516ac781caf4eb18422294cc29fa222090)

We would better add this pr's ut to this pr. If we need integration, we could add it in another single pr.

File: deploy/kubernetes/operator/api/uniffle/v1alpha1/remoteshuffleservice_types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (crain-cn) [reply deploy/kubernetes/operator/api/uniffle/v1alpha1/remoteshuffleservice_types.go:224]: > > > @crain-cn Could you add some uts for this pr?
> > > Already written.
> > > #641 [0cfc745](https://github.com/apache/incubator-uniffle/pull/641/commits/0cfc74516ac781caf4eb18422294cc29fa222090)
> 
> We would better add this pr's ut to this pr. If we need integration, we could add it in another single pr.

PRO asked me to add the affinity field, so I added another pr.  #641

**Comment by crain-cn** [comment]
> > > @crain-cn Could you add some uts for this pr?
> > > Already written.
> > > #641 [0cfc745](https://github.com/apache/incubator-uniffle/pull/641/commits/0cfc74516ac781caf4eb18422294cc29fa222090)
> 
> We would better add this pr's ut to this pr. If we need integration, we could add it in another single pr.

PRO asked me to add an affinity field, so I added another pr. #641

**Comment by crain-cn** [comment]
> Then I close #641

**Pull request closed without merging**